pkg/segment/search/segsearch.go

```diff
@@ -489,42 +490,63 @@ func rawSearchSingleSPQMR(multiReader *segread.MultiColSegmentReader, req *struc
+			utils.And, queryMetrics, qid, allSearchResults, nodeRes, blockSummaries, timeRange)
+		mergeSegmentSearchStatus(searchRes, andSearchRes, utils.And)
 	}
 
 	if node.OrSearchConditions != nil {
-		applyRawSearchToConditions(node.OrSearchConditions, searchReq, segmentSearch, allBlockSearchHelpers,
-			utils.Or, queryMetrics, qid, allSearchResults, nodeRes)
+		orSearchRes := applyRawSearchToConditions(node.OrSearchConditions, searchReq, allBlockSearchHelpers,
+			utils.Or, queryMetrics, qid, allSearchResults, nodeRes, blockSummaries, timeRange)
+		mergeSegmentSearchStatus(searchRes, orSearchRes, utils.And)
 	}
 
 	if node.ExclusionSearchConditions != nil {
-		applyRawSearchToConditions(node.ExclusionSearchConditions, searchReq, segmentSearch, allBlockSearchHelpers,
-			utils.Exclusion, queryMetrics, qid, allSearchResults, nodeRes)
+		notSearchRes := applyRawSearchToConditions(node.ExclusionSearchConditions, searchReq, allBlockSearchHelpers,
+			utils.Exclusion, queryMetrics, qid, allSearchResults, nodeRes, blockSummaries, timeRange)
+		mergeSegmentSearchStatus(searchRes, notSearchRes, utils.And)
 	}
+
+	return searchRes
 }
 
-func applyRawSearchToConditions(cond *structs.SearchCondition, searchReq *structs.SegmentSearchRequest, segmentSearch *SegmentSearchStatus,
+func applyRawSearchToConditions(cond *structs.SearchCondition, searchReq *structs.SegmentSearchRequest,
 	allBlockSearchHelpers []*structs.BlockSearchHelper, op utils.LogicalOperator, queryMetrics *structs.QueryProcessingMetrics, qid uint64,
-	allSearchResults *segresults.SearchResults, nodeRes *structs.NodeResult) {
+	allSearchResults *segresults.SearchResults, nodeRes *structs.NodeResult, blockSummaries []*structs.BlockSummary, timeRange *dtu.TimeRange) *SegmentSearchStatus {
+
+	searchRes := InitBlocksToSearch(searchReq, blockSummaries, allSearchResults, timeRange)
 
 	if cond.SearchNode != nil {
 		for _, sNode := range cond.SearchNode {
-			executeRawSearchOnNode(sNode, searchReq, segmentSearch, allBlockSearchHelpers, queryMetrics,
-				qid, allSearchResults, nodeRes)
+			nodeSearchRes := executeRawSearchOnNode(sNode, searchReq, allBlockSearchHelpers, queryMetrics,
+				qid, allSearchResults, nodeRes, blockSummaries, timeRange)
+			mergeSegmentSearchStatus(searchRes, nodeSearchRes, op)
 		}
 	}
 	if cond.SearchQueries != nil {
 		for _, query := range cond.SearchQueries {
-			RawSearchSingleQuery(query, searchReq, segmentSearch, allBlockSearchHelpers, op, queryMetrics,
+			RawSearchSingleQuery(query, searchReq, searchRes, allBlockSearchHelpers, op, queryMetrics,
 				qid, allSearchResults, nodeRes)
 		}
 	}
+
+	return searchRes
+}
+
+func mergeSegmentSearchStatus(baseSearch *SegmentSearchStatus, searchToMerge *SegmentSearchStatus, op utils.LogicalOperator) {
+	if searchToMerge == nil {
+		return
+	}
+	for blkNum, blkStatus := range searchToMerge.AllBlockStatus {
+		_ = baseSearch.updateMatchedRecords(blkNum, blkStatus.allRecords, op)
+	}
```

Comment:
Why are you ignoring the `error` type that's returned? We probably want to log it

Reply:
sorry missed that. Addressed in the latest commit.